### ISSUE TITLE
cannot zfs send/receive from solaris10 zfs to zfsonlinux

### ISSUE THREAD

**Issue by abrust-ucsd** [zfsonlinux/zfs#5985]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       |  CentOS
Distribution Version    | 7.3
Linux Kernel                 | 3.10.0-514.10.2.el7.x86_64
Architecture                 | x86_64
ZFS Version                  | 0.6.5.9-1
SPL Version                  | 0.6.5.9-1

### Describe the problem you're observing
Sending a zfs snapshot from Solaris10 zpool version 28 to CentOS zfsonlinux, results in this error:
```
cannot receive: stream has unsupported feature, feature flags = 24
```

Both systems are using ZFS filesystem version 5.

I noticed I can send a snapshot that is using an older zpool version (22) without issue.

Am I wrong, or should this be supported (as long as the _source_ zpool version is <= 28)?

### Describe how to reproduce the problem
create a zpool on Solaris10 with version 28
```
zpool create -o version=28 tank /dev/rdsk/c2t2d0
zfs create tank/foobar
zfs snapshot tank/foobar@today
zfs send tank/foobar@today | ssh <remote CentOS host> zfs recv backups
```
### Include any warning/errors/backtraces from the system logs
zpool upgrade confirms the version (on Solaris10):
```
bash-4.3# zpool upgrade
This system is currently running ZFS pool version 32.

The following pools are out of date, and can be upgraded.  After being
upgraded, these pools will no longer be accessible by older software versions.

VER  POOL
---  ------------
28   tank
```

A zsteamdump shows the `features = 24`
```
bash-4.3# zfs send tank/foobar@today > zdump
bash-4.3# zstreamdump < zdump
BEGIN record
        hdrtype = 1
        features = 24
        magic = 2f5bacbac
        creation_time = 58e8080c
        type = 2
        flags = 0x0
        toguid = a7d2ea85b53dbf53
        fromguid = 0
        toname = tank/foobar@today
END checksum = 172c53b8291/2081c5776c1aa9/c37328d967f8e11f/3f3f1fd8f7b7f590
SUMMARY:
        Total DRR_BEGIN records = 1
        Total DRR_END records = 1
        Total DRR_OBJECT records = 7
        Total DRR_FREEOBJECTS records = 3
        Total DRR_WRITE records = 8
        Total DRR_FREE records = 12
        Total DRR_SPILL records = 0
        Total records = 32
        Total write size = 37376 (0x9200)
        Total stream length = 47696 (0xba50)
```

**Comment by kpande** [comment]
http://zfs-discuss.zfsonlinux.narkive.com/LMp4Yqys/heads-up-solaris-10-update-11-zfs-send-incompatible

Not a bug, cannot be fixed by us, sorry..

**Comment by kpande** [comment]
@behlendorf there is a fix here https://github.com/zfs-osx/zfs/commit/521ea8da0d611c4175ac0a37366527ef087e8297

though I'm not sure the long term repercussions.

**Comment by lundman** [comment]
I have used this fix a few times personally, even compiled it on OmniOS to be able to receive snapshots there. But there is no guarantee that something could be bad under the hood after receiving. Luckily we rarely use ACLs so it has worked out for us.

**Comment by ryao** [comment]
@kpande And that is why it was never merged. :/

@abrust-ucsd While none of us can prove it, it is as if Oracle intentionally broke this. We can try ignoring the flag and that seems to work, but it is potentially unsafe because we do not know what they changed.

**Comment by abrust-ucsd** [comment]
Thanks @ryao.  It appears as though I'm able to `zpool export` a solaris zpool and then `zpool import` to ZFS on Linux without issue.  I believe I will take this path and phase out our Solaris systems completely.

**Comment by ioletri** [comment]
@lundman Could you please explain what fix did you talking about?
Thanks for advance.

**Comment by lundman** [comment]
Ah been a while now. 
https://zfs-discuss.zfsonlinux.narkive.com/LMp4Yqys/heads-up-solaris-10-update-11-zfs-send-incompatible#post5

So for us to send Solaris 10 -> OmniOS, I compiled my own `libzfs` where I commented out the exit:

https://github.com/omniti-labs/illumos-omnios/blob/master/usr/src/lib/libzfs/common/libzfs_sendrecv.c#L3681

Then copied my special libzfs.so over to destination and ran:

`LD_PRELOAD=/var/tmp/patched-libzfs.so zfs recv ....`

We never came across the "new type" in any of our Solaris 10 streams. 

It is probably too hacky for OpenZFS to include, but it would be handy to have an advanced tunable which lets you ignore the issue and "hope for the best".
